src/com/amazon/ionschema/internal/constraint/Regex.kt

```diff
@@ -21,55 +21,208 @@ import com.amazon.ion.IonValue
+
+            ch = si.next()
+        } while (ch != null)
+
+        return Pattern.compile(sb.toString(), flags)
+    }
+
+    private fun parseCharacterClass(si: StringIterator, sb: StringBuilder) {
+        var ch = si.next()
+        sb.append(ch)
+
+        if (ch == '^') {
+            ch = si.next()
+            sb.append(ch)
+        }
+
+        ch = si.next()
+        if (ch == '\\') {      // block use of Pattern's "[\p..." constructs
+            error(si, "invalid character '$ch' in character class")
+        }
+        sb.append(ch)
+
+        if (ch == '-') {       // it's a character range
+            ch = si.next()
+            sb.append(ch)
+
+            ch = si.next()
+            when (ch) {
+                ']' -> sb.append(ch)
+                null -> error(si, "unexpected end of string")
+                else -> error(si, "unexpected character '$ch'")
+            }
+
+        } else {               // it's a character class
+            var complete = false
+            ch = si.next()
+            while (ch != null && !complete) {
+                sb.append(ch)
+                if (ch == ']') {
+                    complete = true
+                } else {
+                    ch = si.next()
+                }
+            }
+
+            if (!complete) {
+                error(si, "character class missing ']'")
+            }
+        }
+    }
+
+    private fun parseQuantifier(si: StringIterator, sb: StringBuilder) {
+        val initialLength = sb.length
+        var ch = si.peek()
+        when (ch) {
+            '?', '*', '+' -> {
+                ch = si.next()
+                sb.append(ch)
+            }
+            '{' -> {
+                ch = si.next()
+                sb.append(ch)
+                var complete = false
+                do {
+                    ch = si.next()
+                    when (ch) {
+                        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ',' -> sb.append(ch)
+                        '}' -> {
+                            sb.append(ch)
+                            complete = true
+                        }
+                        null -> {}
+                        else -> error(si, "invalid character '$ch'")
+                    }
+                } while (ch != null && !complete)
+
+                if (!complete) {
+                    error(si, "range quantifier missing '}'")
+                }
+            }
+        }
+
+        if (sb.length > initialLength && ch != null) {
+            ch = si.peek()
+            when (ch) {
+                '?' -> error(si, "invalid character '$ch'")
+                '+' -> error(si, "invalid character '$ch'")
+            }
+        }
+    }
+
+    private fun error(si: StringIterator, message: String): Unit =
+        throw InvalidSchemaException("$message in regex '$si' at offset ${si.index}")
+}
+
+private class StringIterator(private val s: String) {
```

Comment:
probably more memory wasteful but did we consider `ArrayDequeue<Character>`?

Reply:
Hadn't considered it, but `ArrayDeque` is missing a way to determine the current index (well, i suppose queue length could be used for this).  Seems awkward and wasteful; sticking with `StringIterator`.